### PR TITLE
Migrate jsx files that affect run/task selection to tsx

### PR DESCRIPTION
**Pull request by bbovenzi**
In the Grid view, selecting a task or run can be very slow for large DAGs. This is because of some slow querying. I am about to fix that. But I figured it would be safer to make a big change if all the related files were in typescript so issues are caught early.

Partial solution to: https://github.com/apache/airflow/issues/24350

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/www/static/js/grid/types/index.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+type RunState = 'success' | 'running' | 'queued' | 'failed' | 'no_status' | '';
```
Comment (ashb) [inline airflow/www/static/js/grid/types/index.ts]: Do we want both no_status and empty string here?

Comment (ashb) [reply airflow/www/static/js/grid/types/index.ts]: And why empty string and not null?

File: airflow/www/static/js/grid/types/index.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+interface DagRun {
+  runId: string;
+  runType: 'manual' | 'backfill' | 'scheduled';
+  state: RunState;
+  executionDate: string;
+  dataIntervalStart: string;
+  dataIntervalEnd: string;
+  startDate: string;
+  endDate: string;
```
Comment (ashb) [inline airflow/www/static/js/grid/types/index.ts]:
```suggestion
  startDate: string | null;
  endDate: string | null;
```

I think, as these can be empty if the run hasn't started yet.

File: airflow/www/static/js/grid/types/index.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+interface GridTaskInstance {
```
Comment (ashb) [inline airflow/www/static/js/grid/types/index.ts]: Why the `Grid` prefix?

File: airflow/www/static/js/grid/types/index.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  runId: string;
+  taskId: string;
+  startDate: string | null;
+  endDate: string | null;
+  state: TaskState;
+  mappedStates?: {
+    [key: string]: number;
```
Comment (ashb) [inline airflow/www/static/js/grid/types/index.ts:54]:
```suggestion
    [key: TaskState]: number;
```

If that is allowed

**Review by ashb — APPROVED** [review]
Generally looks good -- few small points about the types though.

File: airflow/www/static/js/grid/types/index.ts (hunk 4 above)
Comment (bbovenzi) [reply airflow/www/static/js/grid/types/index.ts:54]: Unfortunately not to my knowledge.

File: airflow/www/static/js/grid/types/index.ts (hunk 1 above)
Comment (bbovenzi) [reply airflow/www/static/js/grid/types/index.ts]: Right, DagRun is only the first four. a Task Instance can have null, fixed.

**Review by ashb — APPROVED** [review]

File: airflow/www/static/js/grid/components/InstanceTooltip.tsx [hunk 5]
```diff
@@ -23,25 +23,33 @@ import { Box, Text } from '@chakra-ui/react';
...
 import Time from './Time';
+import type { TaskInstance, Task } from '../types';
 
-const InstanceTooltip = ({
+interface Props {
+  group: Task;
+  instance: TaskInstance;
+}
+
+const InstanceTooltip: React.FC<Props> = ({
```
Comment (pierrejeambrun) [inline airflow/www/static/js/grid/components/InstanceTooltip.tsx]: Any reason we favor React.FC over:
```
const MyComponent  = ({}: Props): JSX.Element => {}
```

React.FC was removed in CRA:
https://github.com/facebook/create-react-app/pull/8177

React.FC seems discouraged:
https://github.com/typescript-cheatsheets/react#function-components

File: airflow/www/static/js/grid/types/index.ts (hunk 4 above)
Comment (pierrejeambrun) [reply airflow/www/static/js/grid/types/index.ts:54]: You can do something like:
```
[key in TaskState]: number;
```
But the `| null` is an issue here. :(

File: airflow/www/static/js/grid/components/InstanceTooltip.tsx (hunk 5 above)
Comment (bbovenzi) [reply airflow/www/static/js/grid/components/InstanceTooltip.tsx]: Interesting. You're right. As of React 18, it doesn't make sense anymore. I'll update that.

**Review by pierrejeambrun — APPROVED** [review]

**Pull request merged**